### ISSUE TITLE
Boot using Stratis

### ISSUE THREAD

**Issue by agrover** [stratis-storage/stratisd#635]
In an inird environment, we need to somehow read our metadata and activate pools. There is no D-Bus present, so we need to allow for that.

We could create a special mode in stratisd to do this. Another option would be to create another executable akin to how iscsi works with "iscsistart", that sets up pools and then exits.

We also need to find out how to ask Dracut to include our binary in the initrd in the first place. Finally, in the main running phase of the OS, we need to ask systemd to start stratisd if Stratis-initialized blockdevs are present.

**Comment by Conan-Kudo** [comment]
@agrover For an example of how to include binaries with dracut, you can see this: https://github.com/datto/dattobd/blob/master/dist/initramfs/dracut/module-setup.sh

**Comment by agrover** [comment]
Moving to 1.0.

**Comment by tasleson** [comment]
Depends on #831

**Comment by agrover** [comment]
that's it? :smiley: lgtm.

tangential q: so what's the mechanism dracut uses to determine if it needs to put something in the initrd or not?

**Comment by tasleson** [comment]
@agrover Well apparently it's not that simple, I'm hanging just before we pivot with systemd spitting out a message: `...waiting for device dev-stratis-pool-fs.device`

**Comment by tasleson** [comment]
> .. so what's the mechanism dracut uses to determine if it needs to put something in the initrd or not?

I believe this can be accomplished by editing the dracut config file or using `--add` on the command line.   Otherwise the `module-setup.sh` will denote that it needs to be included if the user space binaries are present.

**Comment by tasleson** [comment]
Udev db when running `stratisd-init` in the initramfs for the DM device housing the FS we want to mount, the important part is the `SYSTEMD_READY=0` which is causing systemd to unmount the FS after it is mounted.

```
P: /devices/virtual/block/dm-5
N: dm-5
E: DEVNAME=/dev/dm-5
E: DEVPATH=/devices/virtual/block/dm-5
E: DEVTYPE=disk
E: DM_UDEV_DISABLE_DISK_RULES_FLAG=1
E: DM_UDEV_DISABLE_OTHER_RULES_FLAG=1
E: DM_UDEV_DISABLE_SUBSYSTEM_RULES_FLAG=1
E: DM_UDEV_RULES_VSN=2
E: MAJOR=253
E: MINOR=5
E: SUBSYSTEM=block
E: SYSTEMD_READY=0
E: TAGS=:systemd:
E: USEC_INITIALIZED=106212957
```

If I do `dmsetup remove_all` && `/usr/sbin/stratisd-init --debug` && `sleep 5` && `killadd stratisd-init` and query the db again I see:

```
P: /devices/virtual/block/dm-5
N: dm-5
S: mapper/stratis-1-a7695327630c400f9390b3f8cfab54e9-thin-fs-c1c264b515a14dabb79bc219e0e8c25a
E: DEVLINKS=/dev/mapper/stratis-1-a7695327630c400f9390b3f8cfab54e9-thin-fs-c1c264b515a14dabb79bc219e0e8c25a
E: DEVNAME=/dev/dm-5
E: DEVPATH=/devices/virtual/block/dm-5
E: DEVTYPE=disk
E: DM_NAME=stratis-1-a7695327630c400f9390b3f8cfab54e9-thin-fs-c1c264b515a14dabb79bc219e0e8c25a
E: DM_SUSPENDED=0
E: DM_UDEV_DISABLE_DISK_RULES_FLAG=1
E: DM_UDEV_DISABLE_OTHER_RULES_FLAG=1
E: DM_UDEV_DISABLE_SUBSYSTEM_RULES_FLAG=1
E: DM_UDEV_RULES_VSN=2
E: DM_UUID=stratis-1-a7695327630c400f9390b3f8cfab54e9-thin-fs-c1c264b515a14dabb79bc219e0e8c25a
E: MAJOR=253
E: MINOR=5
E: SUBSYSTEM=block
E: TAGS=:systemd:
E: USEC_INITIALIZED=1693409266
```
And the FS on the block device can be mounted.

This is all pointing to working out the udev device rules for stratis devices which is going to move us to work on the dm cookie values.

**Comment by cawilliamson** [comment]
Hi folks!

I'm experimenting with Stratis and am wondering how you're ending up with stratisd-init ?

Is that just stratisd compiled without dbus? It would be great to get this working.

**Comment by tasleson** [comment]
> I'm experimenting with Stratis and am wondering how you're ending up with stratisd-init ?

@cawilliamson Yes  the default cargo build does not have dbus support, I'm manually doing a rename of the resultant built binary.  Maybe we figure out how to do this automatically?  To enable dbus you need to do `cargo build --features dbus_enabled`

**Comment by tasleson** [comment]
Success! using the existing https://github.com/tasleson/dracut/tree/stratis_support 

Notes on how I got this to work (for myself and others)
1. Ensure you have `/usr/libexec/stratisd` and `/usr/sbin/stratisd-init` and all the other needed bits that are included in the install rpms.  Only one I believe missing is stratisd-init at the moment
2. Copy the directory and all the files in the above repo to `/usr/lib/dracut/modules.d/90stratis`
3. Hack 10-dm.rules to ensure that we get `/dev/disk/by-*` created (see below)
4. Build a new initramfs, aka `# dracut --add stratis --force`
   * Make sure files are included in initramfs file (see below)
5. Duplicate your root fs to a stratis backed xfs file system.  This ended up being a time sink for me.  I initially tried to rsync rootfs and ended up in an infinite loop, i ended up doing:
   * Shut down VM
   * Duplicate virtual disk for rootfs
   * Boot VM
   * Change the UUID of the FS for the duplicated vdisk 
   * dd the duplicated disk to the stratis backed FS
   * Change the UUID for the stratis backed XFS
   * Grow the XFS filesystem to match what stratis would be expecting
6. Mount stratis backed root fs and update `etc/fstab` to reference new UUID
7. Edit /boot/grub2/grub.cfg and add a new menu entry which references new root UUID
8. Re-boot, select new entry and cross fingers :-)

```
# lsinitrd | grep stratis
Arguments: --add 'stratis' --force
stratis
-rwxr-xr-x   1 root     root          119 Mar  5 03:18 usr/lib/dracut/hooks/cleanup/25-stratisd-stop.sh
-rwxr-xr-x   1 root     root           51 Mar  5 03:18 usr/lib/dracut/hooks/cmdline/25-stratisd-start.sh
-rwxr-xr-x   1 root     root      3455968 Mar  5 03:18 usr/sbin/stratisd-init
```


Some of this diff is debug
``` diff
# diff /usr/lib/udev/rules.d/10-dm.rules /usr/lib/udev/rules.d/10-dm.rules.original 
65c65
< IMPORT{db}="DM_UDEV_ENABLE_DISK_RULES_FLAG"
---
> IMPORT{db}="DM_UDEV_DISABLE_DISK_RULES_FLAG"
72d71
< ENV{DM_TONY_NO_COOKIE_DEBUG}="1"
146,147d144
< ENV{DM_NAME}="$attr{dm/name}"
< ENV{DM_TONY_DM_DISABLE_DEBUG}="1"
149,150c146,147
< ENV{DM_UDEV_DISABLE_DISK_RULES_FLAG}="0"
< ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}="0"
---
> ENV{DM_UDEV_DISABLE_DISK_RULES_FLAG}="1"
> ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}="1"

```

**Comment by tasleson** [comment]
Broken by: https://github.com/stratis-storage/stratisd/pull/986

**Comment by agrover** [comment]
I opened #987, should I close it as a dupe of this?

**Comment by tasleson** [comment]
@agrover Leave the new issue open, this one is going to be getting closed soon once the cookie handling is in place.

**Comment by tasleson** [comment]
Resolved with https://github.com/stratis-storage/devicemapper-rs/pull/334

**Comment by shibumi** [comment]
Little question:
Does this work already? did somebody write up how to boot from a stratis provisioned filesystem? I am talking about using stratis for things like root. Would this work or is this not the purpose of stratis?

**Comment by tasleson** [comment]
@shibumi Yes this should work for root fs.  No write up has been done other than what is mentioned in this issue already.

**Comment by q2dg** [comment]
I raise the bet: could it be possible to install a new fresh (RedHat/Fedora) system (via Anaconda) into a Stratis pool as a root filesystem like it can be done nowadays into a LVM/RAID/iSCSI device? If so, how could it be done? (although it implies manual terminal commands execution, never mind)
Thanks!!
